omegaconf/base.py

```diff
@@ -377,28 +400,63 @@ def _resolve_interpolation(
         elif value_kind == ValueKind.STR_INTERPOLATION:
             value = _get_value(value)
             assert isinstance(value, str)
-            orig = value
-            new = ""
-            last_index = 0
-            for match in match_list:
-                new_val = self._resolve_simple_interpolation(
+            try:
+                # NB: `match_list` is ignored here as complex interpolations may be
+                # nested, the string will be re-parsed within `_evaluate_complex()`.
+                new = self._evaluate_complex(
+                    value,
                     key=key,
-                    inter_type=match.group(1),
-                    inter_key=match.group(2),
                     throw_on_missing=throw_on_missing,
                     throw_on_resolution_failure=throw_on_resolution_failure,
                 )
-                # if failed to resolve, return None for the whole thing.
-                if new_val is None:
-                    return None
-                new += orig[last_index : match.start(0)] + str(new_val)
-                last_index = match.end(0)
-
-            new += orig[last_index:]
+            except ParseError:
+                return None
             return StringNode(value=new, key=key)
         else:
             assert False
 
+    def _evaluate_simple(self, value: str, **kw: Any) -> Optional["Node"]:
+        """Evaluate a simple interpolation"""
+        value_kind, match_list = get_value_kind(value=value, return_match_list=True)
```

Comment:
Can you reuse the previous parsing results here instead of re-parse?

Reply:
I could try but the problem is that the current parsing results are "incomplete", for instance if the string is `${f.${c}}` then it only matches `${c}`. So there could be two options:
1. Re-use the matches that have been found before: should be relatively easy to implement, but the code would be harder to read => not sure it's worth it
2. Improve the matching logic in `get_value_kind()` to fully cover all matches: this seems tricky, at least if we want to keep the current RE-based approach (https://stackoverflow.com/questions/133601/can-regular-expressions-be-used-to-match-nested-patterns)

Reply:
Okay, don't worry about it then.
Once we have a proper grammar this will be a million time easier.
in fact, check this out:
https://github.com/facebookresearch/hydra/blob/master/hydra/grammar/Override.g4#L78-L84

I think it will need to be changed to support complex interpolations from the command line.
In fact, I think it will need to be promoted from a token to a full fledged rule.

Reply:
Yes a grammar will make things cleaner and more consistent. At least this PR should still enable some useful use cases in the meantime.

I ran into another issue related to the regex logic in `get_value_kind()`, which prevented me from fixing the inconsistency between `${a` (no error) and `${a}_${a ` (parsing error), which I had pointed out in point 3 above. It'd be a bit long to explain it in details, and annoying to fix, so I decided to give up on it for now since the grammar should fix it one way or another (= either always allowing it or always throwing an error).

Reply:
Please explain the problem more and how a user would hit this.
I like this change but I want to make sure it does not opening up significant issues we are not ready to deal with.
It will be MANY months before OmegaConf 2.1.

Reply:
With both the current master and this PR, you can use:
```
a: 1
b: ${a
```
and it will run just fine, with `conf.b` being equal to `"${a"`.

However, with this PR, if you do:
```
a: 1
b: x_${a
```
then trying to access `conf.b` will trigger a parsing error. This is because by adding `x_` as prefix, it's not a simple interpolation anymore, so it goes through my new parsing code that ensures that all opening braces are properly closed.

Ideally I would like the first case to also trigger an error. I don't think anyone should be relying on the fact that right now it "works". But I don't see an easy way to make it happen.

So this is not really adding a problem, it's more that I'm unable to fix an existing problem :)

Reply:
This is fine. It's good that you are noticing these differences but don't try too hard to preserve 100% compatibility.
some small differences around error cases are fine.